static/js/settings_bots.js

```diff
@@ -422,112 +421,121 @@ export function set_up() {
+            avatar_widget.clear();
+
+            form.validate({
+                errorClass: "text-error",
+                success() {
+                    errors.hide();
+                },
+                submitHandler() {
+                    const bot_id = Number.parseInt(form.attr("data-user-id"), 10);
+                    const type = form.attr("data-type");
+
+                    const full_name = form.find(".edit_bot_name").val();
+                    const bot_owner_id = owner_widget.value();
+                    const file_input = $(".edit_bot_form").find(".edit_bot_avatar_file_input");
+
+                    const formData = new FormData();
+                    formData.append("csrfmiddlewaretoken", csrf_token);
+                    formData.append("full_name", full_name);
+                    formData.append("bot_owner_id", bot_owner_id);
+
+                    if (type === OUTGOING_WEBHOOK_BOT_TYPE) {
+                        const service_payload_url = $("#edit_service_base_url").val();
+                        const service_interface = $("#edit_service_interface :selected").val();
+                        formData.append("service_payload_url", JSON.stringify(service_payload_url));
+                        formData.append("service_interface", service_interface);
+                    } else if (type === EMBEDDED_BOT_TYPE) {
+                        const config_data = {};
+                        $("#config_edit_inputbox input").each(function () {
+                            config_data[$(this).attr("name")] = $(this).val();
+                        });
+                        formData.append("config_data", JSON.stringify(config_data));
+                    }
+                    for (const [i, file] of Array.prototype.entries.call(file_input[0].files)) {
+                        formData.append("file-" + i, file);
+                    }
+                    channel.patch({
+                        url: "/json/bots/" + encodeURIComponent(bot_id),
+                        data: formData,
+                        cache: false,
+                        processData: false,
+                        contentType: false,
+                        success(data) {
+                            avatar_widget.clear();
+                            if (data.avatar_url) {
+                                // Note that the avatar_url won't actually change on the backend
+                                // when the user had a previous uploaded avatar.  Only the content
+                                // changes, so we version it to get an uncached copy.
+                                image_version += 1;
+                                image
+                                    .find("img")
+                                    .attr(
+                                        "src",
+                                        data.avatar_url + "&v=" + image_version.toString(),
+                                    );
+                            }
```

Comment:
In theory, shouldn't this live-update work be done in the `server_events_dispatch` code path for the avatar change, rather than here?  If so let's open an issue for that.